File: apps/ios/Loonext/Features/Compose/Drafts.swift
import Foundation

/// Client-side composer drafts, one per conversation (SPEC: the server keeps
/// NO drafts — restore-on-failure and cross-open persistence are purely ours).
/// Text only: staged photos/files reference transient picker grants that do
/// not survive the process, so persisting them would restore dead chips.
@MainActor
final class ComposerDrafts {
    /// The new-conversation screen's draft rides a fixed slot.
    static let newConversation = "new"

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    private func key(_ conversationId: String) -> String {
        "composer-draft:\(conversationId)"
    }

    func load(_ conversationId: String) -> String {
        defaults.string(forKey: key(conversationId)) ?? ""
    }

    func save(_ conversationId: String, text: String) {
        if text.isBlank {
            defaults.removeObject(forKey: key(conversationId))
        } else {
            defaults.set(text, forKey: key(conversationId))
        }
    }

    func clear(_ conversationId: String) {
        defaults.removeObject(forKey: key(conversationId))
        defaults.removeObject(forKey: mentionKey(conversationId))
    }

    // MARK: - Note mentions
    //
    // The teammates named on a note draft ride WITH the text. Persisting only
    // the words restored a draft that still read "@Sam" and notified nobody,
    // which is worse than losing the draft: the note on screen was evidence of
    // something that would not happen.
    //
    // JSON under a separate key, so a draft written by an older build still
    // loads and a value we cannot decode costs the picks rather than the draft.

    private func mentionKey(_ conversationId: String) -> String {
        "composer-draft-mentions:\(conversationId)"
    }

    func loadMentions(_ conversationId: String) -> [PickedMention] {
        guard let data = defaults.data(forKey: mentionKey(conversationId)) else { return [] }
        return (try? JSONDecoder().decode([PickedMention].self, from: data)) ?? []
    }

    func saveMentions(_ conversationId: String, mentions: [PickedMention]) {
        if mentions.isEmpty {
            defaults.removeObject(forKey: mentionKey(conversationId))
        } else if let data = try? JSONEncoder().encode(mentions) {
            defaults.set(data, forKey: mentionKey(conversationId))
        }
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/compose/Drafts.kt
package com.loonext.android.features.compose

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.loonext.android.features.thread.PickedMention
import kotlinx.coroutines.flow.first
import kotlinx.serialization.json.Json

private val Context.composerDraftStore by preferencesDataStore(name = "composer-drafts")

/**
 * Client-side composer drafts, one per conversation (SPEC: the server keeps NO
 * drafts — restore-on-failure and cross-open persistence are purely ours).
 * Text only: staged photos/files are content URIs whose read permission does
 * not survive the process, so persisting them would restore dead chips.
 */
class ComposerDrafts(private val context: Context) {

    private fun key(conversationId: String) = stringPreferencesKey("draft:$conversationId")

    /** The new-conversation screen's draft rides a fixed slot. */
    companion object {
        const val NEW_CONVERSATION = "new"
    }

    suspend fun load(conversationId: String): String =
        context.composerDraftStore.data.first()[key(conversationId)].orEmpty()

    suspend fun save(conversationId: String, text: String) {
        context.composerDraftStore.edit { prefs ->
            if (text.isBlank()) prefs.remove(key(conversationId))
            else prefs[key(conversationId)] = text
        }
    }

    suspend fun clear(conversationId: String) {
        context.composerDraftStore.edit {
            it.remove(key(conversationId))
            it.remove(mentionKey(conversationId))
        }
    }

    // --- Note mentions ------------------------------------------------------
    //
    // The teammates named on a note draft ride WITH the text. Persisting only
    // the words restored a draft that still read "@Sam" and notified nobody,
    // which is worse than losing the draft: the note on screen was evidence of
    // something that would not happen.
    //
    // JSON under a separate key, so a draft written by an older build still
    // loads and a value we cannot parse costs the picks rather than the draft.

    private fun mentionKey(conversationId: String) =
        stringPreferencesKey("draft-mentions:$conversationId")

    suspend fun loadMentions(conversationId: String): List<PickedMention> {
        val raw = context.composerDraftStore.data.first()[mentionKey(conversationId)]
        if (raw.isNullOrEmpty()) return emptyList()
        return runCatching { Json.decodeFromString<List<PickedMention>>(raw) }
            .getOrDefault(emptyList())
    }

    suspend fun saveMentions(conversationId: String, mentions: List<PickedMention>) {
        context.composerDraftStore.edit { prefs ->
            if (mentions.isEmpty()) {
                prefs.remove(mentionKey(conversationId))
            } else {
                prefs[mentionKey(conversationId)] = Json.encodeToString(mentions)
            }
        }
    }
}
